From 8340885dbe2fee30d86084a24cbc78705d40bce0 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Fri, 9 Oct 2020 18:00:11 +0800
Subject: [PATCH 07/94] libweston: Avoid double-destroy of animations

The idle_animation_destroy task should be removed when destroying
animations by the other callers such as handle_animation_view_destroy().

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 libweston/animation.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/libweston/animation.c b/libweston/animation.c
index 22a2e1cfeb..e41831785f 100644
--- a/libweston/animation.c
+++ b/libweston/animation.c
@@ -136,6 +136,7 @@ struct weston_view_animation {
 	weston_view_animation_frame_func_t frame;
 	weston_view_animation_frame_func_t reset;
 	weston_view_animation_done_func_t done;
+	struct wl_event_source *idle_destroy_source;
 	void *data;
 	void *private;
 };
@@ -143,6 +144,9 @@ struct weston_view_animation {
 WL_EXPORT void
 weston_view_animation_destroy(struct weston_view_animation *animation)
 {
+	if (animation->idle_destroy_source)
+		wl_event_source_remove(animation->idle_destroy_source);
+
 	wl_list_remove(&animation->animation.link);
 	wl_list_remove(&animation->listener.link);
 	weston_view_remove_transform(animation->view, &animation->transform);
@@ -207,6 +211,7 @@ idle_animation_destroy(void *data)
 {
 	struct weston_view_animation *animation = data;
 
+	animation->idle_destroy_source = NULL;
 	weston_view_animation_destroy(animation);
 }
 
@@ -223,7 +228,7 @@ weston_view_animation_create(struct weston_view *view,
 	struct weston_compositor *ec = view->surface->compositor;
 	struct wl_event_loop *loop;
 
-	animation = malloc(sizeof *animation);
+	animation = zalloc(sizeof *animation);
 	if (!animation)
 		return NULL;
 
@@ -250,7 +255,9 @@ weston_view_animation_create(struct weston_view *view,
 	} else {
 		wl_list_init(&animation->animation.link);
 		loop = wl_display_get_event_loop(ec->wl_display);
-		wl_event_loop_add_idle(loop, idle_animation_destroy, animation);
+		animation->idle_destroy_source =
+			wl_event_loop_add_idle(loop, idle_animation_destroy,
+					       animation);
 	}
 
 	return animation;
-- 
2.20.1

